components/caption_images/src/main.py

```diff
@@ -22,37 +19,67 @@
 logger = logging.getLogger(__name__)
 
 
-@dask.delayed
-def load(example):
-    bytes = io.BytesIO(example)
-    image = Image.open(bytes).convert("RGB")
-    return image
-
-
-@dask.delayed
-def transform(image, processor, device):
-    inputs = processor(images=image, return_tensors="pt").to(device)
-
-    return inputs
-
-
-@dask.delayed
-def collate(examples):
-    return torch.cat([ex["pixel_values"] for ex in examples])
-
-
-@dask.delayed
-def caption(batch, model, processor, max_new_tokens):
-    logger.info("Generating caption...")
-    generated_ids = model.generate(pixel_values=batch, max_new_tokens=max_new_tokens)
-    generated_captions = processor.batch_decode(generated_ids, skip_special_tokens=True)
-
-    return generated_captions
+def process_image(image: bytes, *, processor: BlipProcessor, device: str) -> torch.Tensor:
+    """
+    Process the image to a tensor.
 
+    Args:
+        image: The input image as a byte string.
+        processor: The processor object for transforming the image.
+        device: The device to move the transformed image to.
+    """
+    def load(img: bytes) -> Image:
+        """Load the bytestring as an image"""
+        bytes_ = io.BytesIO(img)
+        return Image.open(bytes_).convert("RGB")
 
-@dask.delayed
-def flatten(lst):
-    return pd.Series(itertools.chain(*lst))
+    def transform(img: Image) -> BatchEncoding:
+        """
+        Transform the image to a tensor using a clip processor and move it to the specified device.
+        """
+        return processor(images=img, return_tensors="pt").to(device)
+
+    return transform(load(image))["pixel_values"]
+
+
+def caption_image_batch(
+    image_batch: pd.DataFrame,
+    *,
+    model: BlipForConditionalGeneration,
+    processor: BlipProcessor,
+    max_new_tokens: int
+) -> pd.Series:
+    """Caption a batch of images"""
+    input_batch = torch.cat(image_batch.tolist())
+    output_batch = model.generate(pixel_values=input_batch, max_new_tokens=max_new_tokens)
+    captions_batch = processor.batch_decode(output_batch, skip_special_tokens=True)
+
+    return pd.Series(captions_batch, index=image_batch.index)
+
+
+def caption_images(
+    images: pd.Series,
+    *,
+    model: BlipForConditionalGeneration,
+    processor: BlipProcessor,
+    batch_size: int,
+    max_new_tokens: int,
+    device: str,
+):
+    """Caption a pandas series of images"""
+    images = images.apply(process_image, processor=processor, device=device)
```

Comment:
So we're not leveraging parallellism for prepocessing the images?

As this seems to be just Pandas